File: app/src/main/java/by/yahorfralou/plaincalendar/widget/data/calendars/job/CalendarChangeJobService.java
package by.yahorfralou.plaincalendar.widget.data.calendars.job;

import android.app.job.JobParameters;
import android.app.job.JobService;
import android.appwidget.AppWidgetManager;
import android.content.Context;
import android.util.Log;

import java.util.Arrays;

import by.yahorfralou.plaincalendar.widget.R;
import by.yahorfralou.plaincalendar.widget.helper.WidgetHelper;
import by.yahorfralou.plaincalendar.widget.widget.CalendarWidgetProvider;

import static by.yahorfralou.plaincalendar.widget.app.PlainCalendarWidgetApp.LOGCAT;

public class CalendarChangeJobService extends JobService {
    @Override
    public boolean onStartJob(JobParameters params) {
        Context ctx = getApplicationContext();
        Log.i(LOGCAT, "JOB Service#onStartJob");

        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(ctx);
        int[] widgetIds = WidgetHelper.getWidgetIds(ctx, CalendarWidgetProvider.class);
        Log.i(LOGCAT, "Widgets to update in JobService: " + Arrays.toString(widgetIds));

        appWidgetManager.notifyAppWidgetViewDataChanged(widgetIds, R.id.listEvents);

        return true;
    }

    @Override
    public boolean onStopJob(JobParameters params) {
        return true;
    }
}

File: app/src/main/java/by/offvanhooijdonk/plaincalendarv2/widget/data/CalendarChangeJobService.kt
package by.offvanhooijdonk.plaincalendarv2.widget.data

import android.app.job.JobInfo
import android.app.job.JobParameters
import android.app.job.JobScheduler
import android.app.job.JobService
import android.content.ComponentName
import android.content.Context
import android.net.Uri
import android.util.Log
import by.offvanhooijdonk.plaincalendarv2.widget.app.App
import by.offvanhooijdonk.plaincalendarv2.widget.glance.PlainGlanceWidget

class CalendarChangeJobService : JobService() {

    override fun onStartJob(params: JobParameters): Boolean {
        Log.d(App.LOGCAT, "JOB Service#onStartJob")

        PlainGlanceWidget().loadData()

        scheduleCalendarChangeJob(applicationContext)

        return false
    }

    override fun onStopJob(params: JobParameters): Boolean = true

    companion object {
        private const val TRIGGER_MAX_DELAY_TIME: Long = 1000
        private const val JOB_CALENDAR_CHANGE_ID = 1002

        fun scheduleCalendarChangeJob(ctx: Context) {
            Log.d(App.LOGCAT, "Creating Job Schedule.")

            val uriTrigger: Uri = CalendarDataSource.makeEventsObservationUri()
            val builder = JobInfo.Builder(JOB_CALENDAR_CHANGE_ID, ComponentName(ctx, CalendarChangeJobService::class.java))
                .addTriggerContentUri(JobInfo.TriggerContentUri(uriTrigger, JobInfo.TriggerContentUri.FLAG_NOTIFY_FOR_DESCENDANTS))
                .setTriggerContentMaxDelay(TRIGGER_MAX_DELAY_TIME)
            ctx.getSystemService(JobScheduler::class.java)?.schedule(builder.build())
        }

        fun cancelCalendarChangeJob(ctx: Context) {
            ctx.getSystemService(JobScheduler::class.java).cancelAll()
        }
    }
}
